File: packages/types/src/mesh.ts
/**
 * @kite-vpn/types — Mesh network type definitions.
 *
 * Kite 内嵌 Nebula（Slack 开源的 P2P Mesh）作为 sidecar，实现"翻墙 + Mesh
 * 一体化"。本文件定义 Mesh 相关的所有共享类型。
 *
 * 设计原则（per workspace claude.md）：
 * - 禁止滥用 `?` 可选：只在语义上"可能不存在"时用，比如 lastSeenAt 在节点从未
 *   上线过时不存在；name 这种必有字段不用 `?`
 * - 禁止隐藏变量：所有 union / 枚举显式列出可能值
 */

// ---------------------------------------------------------------------------
// Roles & states
// ---------------------------------------------------------------------------

/**
 * Peer 在 Mesh 网络中的角色。
 *
 * - `owner`:  CA 私钥持有者，能签发新证书。每个 Mesh 网络 1 个 owner。
 * - `member`: 普通成员，只能收发流量，不能签证书。
 * - `exit`:   暴露为 exit node，其他 peer 可选择把所有出墙流量经其转发。
 * - `subnet`: 子网路由器，advertise 一个 LAN CIDR（例如家庭 192.168.1.0/24），
 *             让其他 peer 通过它访问那个 LAN。
 * - `lighthouse`: 公网 VPS 上跑的发现服务，不参与数据流量，只帮 peer NAT 穿透。
 *
 * 一个 peer 可以同时是 `member` + `exit` + `subnet`（位掩码语义），但
 * `owner` 和 `lighthouse` 互斥于其他角色。前端用数组表示组合：
 * `roles: ['member', 'exit']`。
 */
export type MeshPeerRole = 'owner' | 'member' | 'exit' | 'subnet' | 'lighthouse'

/**
 * Peer 在线状态 —— 显式枚举，不用 boolean，因为"正在握手"是真实第三态。
 */
export type MeshPeerStatus = 'online' | 'offline' | 'handshaking'

// ---------------------------------------------------------------------------
// Peer
// ---------------------------------------------------------------------------

/**
 * 一个 Mesh 节点。
 *
 * 字段命名规则：camelCase TS 端，Rust 端通过 serde rename 转 snake_case。
 */
export interface MeshPeer {
  /** Stable identifier — 取自证书的 fingerprint (sha256 前 16 hex)。 */
  id: string
  /** 人类可读名称，例：「Mac 笔记本」「家里 NAS」。Owner 设备签发证书时设置。 */
  name: string
  /** 在 Mesh 内的虚拟 IP，例：100.64.0.2。每个 peer 必有，不可选。 */
  meshIp: string
  /** 角色组合，至少 1 个。 */
  roles: MeshPeerRole[]
  /** 当前在线状态。 */
  status: MeshPeerStatus
  /** 最近一次心跳 unix 毫秒时间戳。节点从未上线时为 0（不用 `?` 隐藏空值）。 */
  lastSeenAt: number
  /** 节点公网端点（host:port），NAT 穿透成功后填充。未穿透前为空字符串。 */
  publicEndpoint: string
  /** 证书签发时间 unix 毫秒。用于显示「3 个月前加入」。 */
  enrolledAt: number
  /** 证书过期时间 unix 毫秒。Owner 续签前显示警告。 */
  certExpiresAt: number
  /** subnet 角色时：advertise 的 LAN CIDR，例：「192.168.1.0/24」。其他角色为空字符串。 */
  advertisedSubnet: string
}

// ---------------------------------------------------------------------------
// Network
// ---------------------------------------------------------------------------

/**
 * 一个完整的 Mesh 网络。v1 阶段一个 Kite 实例只能加入 1 个网络（v3 支持多网络）。
 */
export interface MeshNetwork {
  /** 网络 ID — 取自 CA 证书 fingerprint。 */
  id: string
  /** 网络名称，例「家庭网络」，owner 设置。 */
  name: string
  /** CIDR 网段，默认 100.64.0.0/10（RFC 6598）。 */
  cidr: string
  /** Lighthouse 公网地址（host:port），node 通过它做发现 + 穿透 fallback。 */
  lighthouseEndpoint: string
  /** 网络创建时间 unix 毫秒。 */
  createdAt: number
  /** 本设备在该网络里的 peer ID（外键到 MeshPeer.id）。 */
  selfPeerId: string
}

// ---------------------------------------------------------------------------
// Enrollment — 新设备加入流程
// ---------------------------------------------------------------------------

/**
 * Owner 设备生成的一次性 enrollment token，用于让新设备加入网络。
 *
 * 工作流：
 *   1. Owner 在 Kite 点「添加设备」→ 生成 token + 上传加密载荷到 lighthouse
 *   2. 新设备装 Kite → 输入 token → 从 lighthouse 取回签好的证书 + 配置
 *   3. Token 一次性使用，过期作废
 */
export interface MeshEnrollmentToken {
  /** Token 字符串本身，base32 编码，约 24 字符，方便用户手抄 / 二维码扫描。 */
  token: string
  /** 这个 token 是给哪个新 peer 名字签发的，例：「我的 iPhone」。 */
  peerName: string
  /** 预设角色，owner 在签发前决定。 */
  roles: MeshPeerRole[]
  /** 预分配的 mesh IP。 */
  meshIp: string
  /** Unix 毫秒过期时间，默认创建后 10 分钟。 */
  expiresAt: number
}

// ---------------------------------------------------------------------------
// ACL — 哪台设备能访问哪台
// ---------------------------------------------------------------------------

/**
 * 一条 ACL 规则。Owner 在 Kite UI 编辑后，重签所有 peer 证书把规则烧进去。
 *
 * 语义：from 中的任意 peer 可以访问 to 中的任意 peer 的 ports（TCP/UDP/ICMP）。
 *
 * 默认（无 ACL）：所有 peer 可以访问彼此所有端口（最宽松，便于 v1 调试）。
 */
export interface MeshAclRule {
  /** UUID，用户重命名规则时保持稳定。 */
  id: string
  /** 描述性名称，例「手机只能访问 NAS 的 SMB」。 */
  name: string
  /** 源 peer ID 列表，空数组表示「任何 peer」。 */
  from: string[]
  /** 目标 peer ID 列表，空数组表示「任何 peer」。 */
  to: string[]
  /** TCP/UDP 端口列表，例 ['22', '80', '443', '5000-5100']，空数组表示「所有端口」。 */
  ports: string[]
  /** 是否允许 ICMP（ping）。Boolean 显式声明，不用 `?`。 */
  allowIcmp: boolean
}

// ---------------------------------------------------------------------------
// Mesh 引擎运行时状态（IPC engine_get_state 用）
// ---------------------------------------------------------------------------

export type MeshEngineStatus = 'stopped' | 'starting' | 'running' | 'error'

export interface MeshEngineState {
  status: MeshEngineStatus
  /** 当前进程的 PID，未运行时为 0（不用 `?` 隐藏空值）。 */
  pid: number
  /** Nebula 版本号，例「v1.9.5」。从二进制读出，加载失败时为空字符串。 */
  version: string
  /** 错误描述，仅 status='error' 时有内容，其他状态为空字符串。 */
  error: string
}

File: apps/desktop/src-tauri/src/commands/mesh.rs
//! Mesh 网络 IPC（Phase 4，基于 Nebula sidecar）。
//!
//! 这一版是**框架 + 序列化契约**，不真正启动 nebula 进程。等 sidecar 集成完成
//! 后会回填实际逻辑。先把 IPC 形状钉死，让前端可以并行开发。
//!
//! 设计原则（per workspace claude.md）：
//! - 业界最佳实践：用 serde 显式 rename camelCase 跟 TS 端对齐
//! - 禁止隐藏变量：所有 IPC 入参 / 返回值类型显式，不用 serde_json::Value 偷懒
//! - 多写测试：每个公开类型有 round-trip serde 测试
//!
//! 当 nebula sidecar 真正集成后，每个 #[tauri::command] 函数的 body 会调
//! nebula-cert / nebula 子进程；类型签名不变。
//!
//! 详细设计见：
//! ~/.gstack/projects/vpn/fengwenxuan-main-design-20260514-015705-mesh.md

use serde::{Deserialize, Serialize};

use super::IpcResult;

// ─── 类型定义（跟 packages/types/src/mesh.ts 对齐） ─────────────────────────

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MeshPeerRole {
    Owner,
    Member,
    Exit,
    Subnet,
    Lighthouse,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MeshPeerStatus {
    Online,
    Offline,
    Handshaking,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MeshEngineStatus {
    Stopped,
    Starting,
    Running,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MeshPeer {
    pub id: String,
    pub name: String,
    pub mesh_ip: String,
    pub roles: Vec<MeshPeerRole>,
    pub status: MeshPeerStatus,
    pub last_seen_at: i64,
    pub public_endpoint: String,
    pub enrolled_at: i64,
    pub cert_expires_at: i64,
    pub advertised_subnet: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MeshNetwork {
    pub id: String,
    pub name: String,
    pub cidr: String,
    pub lighthouse_endpoint: String,
    pub created_at: i64,
    pub self_peer_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MeshEnrollmentToken {
    pub token: String,
    pub peer_name: String,
    pub roles: Vec<MeshPeerRole>,
    pub mesh_ip: String,
    pub expires_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MeshEngineState {
    pub status: MeshEngineStatus,
    pub pid: u32,
    pub version: String,
    pub error: String,
}

impl MeshEngineState {
    /// Helper: 默认"已停止"状态，所有数值字段显式为 0 / 空字符串，
    /// 符合"禁止滥用 Option 隐藏空值"原则。
    pub fn stopped() -> Self {
        Self {
            status: MeshEngineStatus::Stopped,
            pid: 0,
            version: String::new(),
            error: String::new(),
        }
    }
}

// ─── IPC 命令（v1 framework 阶段：全部返回未实现状态） ─────────────────────

const NOT_IMPLEMENTED: &str = "Mesh sidecar 尚未集成 —— framework only, 等待 nebula 二进制就位";

/// 查询 Mesh 引擎运行状态。
#[tauri::command]
pub async fn mesh_get_engine_state() -> IpcResult<MeshEngineState> {
    IpcResult::ok(MeshEngineState::stopped())
}

/// 启动 Mesh 引擎（前提：本设备已加入某个网络，配置文件已落盘）。
#[tauri::command]
pub async fn mesh_start() -> IpcResult<MeshEngineState> {
    IpcResult::err(NOT_IMPLEMENTED)
}

/// 停止 Mesh 引擎。
#[tauri::command]
pub async fn mesh_stop() -> IpcResult<MeshEngineState> {
    IpcResult::err(NOT_IMPLEMENTED)
}

/// 列出本设备所在网络的所有 peers。
#[tauri::command]
pub async fn mesh_list_peers() -> IpcResult<Vec<MeshPeer>> {
    IpcResult::ok(Vec::new())
}

/// 当前网络元数据。未加入任何网络时返回 None。
#[tauri::command]
pub async fn mesh_get_network() -> IpcResult<Option<MeshNetwork>> {
    IpcResult::ok(None)
}

/// 创建新的 Mesh 网络（首次启动 Kite 用）。
/// 生成 CA + owner 证书 + 默认配置。返回创建好的网络元数据。
#[tauri::command]
pub async fn mesh_create_network(name: String, lighthouse_endpoint: String) -> IpcResult<MeshNetwork> {
    let _ = (name, lighthouse_endpoint);
    IpcResult::err(NOT_IMPLEMENTED)
}

/// Owner 生成一次性 enrollment token，供新设备加入。
#[tauri::command]
pub async fn mesh_generate_enrollment_token(
    peer_name: String,
    roles: Vec<MeshPeerRole>,
) -> IpcResult<MeshEnrollmentToken> {
    let _ = (peer_name, roles);
    IpcResult::err(NOT_IMPLEMENTED)
}

/// 新设备用 token 加入网络。
#[tauri::command]
pub async fn mesh_join_network(token: String) -> IpcResult<MeshNetwork> {
    let _ = token;
    IpcResult::err(NOT_IMPLEMENTED)
}

/// 撤销一个 peer（Owner only）。
#[tauri::command]
pub async fn mesh_revoke_peer(peer_id: String) -> IpcResult<()> {
    let _ = peer_id;
    IpcResult::err(NOT_IMPLEMENTED)
}

// ─── 单元测试（rule 4：多写测试 case） ──────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn engine_status_serializes_lowercase() {
        let s = serde_json::to_string(&MeshEngineStatus::Stopped).unwrap();
        assert_eq!(s, "\"stopped\"");
        let s = serde_json::to_string(&MeshEngineStatus::Running).unwrap();
        assert_eq!(s, "\"running\"");
    }

    #[test]
    fn peer_role_serializes_lowercase() {
        let s = serde_json::to_string(&MeshPeerRole::Owner).unwrap();
        assert_eq!(s, "\"owner\"");
        let s = serde_json::to_string(&MeshPeerRole::Lighthouse).unwrap();
        assert_eq!(s, "\"lighthouse\"");
    }

    #[test]
    fn peer_round_trip_camelcase() {
        let peer = MeshPeer {
            id: "abc123".into(),
            name: "Mac 笔记本".into(),
            mesh_ip: "100.64.0.2".into(),
            roles: vec![MeshPeerRole::Member, MeshPeerRole::Subnet],
            status: MeshPeerStatus::Online,
            last_seen_at: 1_700_000_000_000,
            public_endpoint: "203.0.113.5:4242".into(),
            enrolled_at: 1_690_000_000_000,
            cert_expires_at: 1_720_000_000_000,
            advertised_subnet: "192.168.1.0/24".into(),
        };
        let json = serde_json::to_string(&peer).unwrap();
        // 确认 camelCase rename 生效
        assert!(json.contains("\"meshIp\":\"100.64.0.2\""));
        assert!(json.contains("\"lastSeenAt\":1700000000000"));
        assert!(json.contains("\"advertisedSubnet\":\"192.168.1.0/24\""));
        // 反序列化往返
        let restored: MeshPeer = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.id, peer.id);
        assert_eq!(restored.mesh_ip, peer.mesh_ip);
        assert_eq!(restored.roles.len(), 2);
    }

    #[test]
    fn engine_state_stopped_has_zero_pid_and_empty_strings() {
        let s = MeshEngineState::stopped();
        assert_eq!(s.pid, 0);
        assert_eq!(s.version, "");
        assert_eq!(s.error, "");
        assert!(matches!(s.status, MeshEngineStatus::Stopped));
    }

    #[test]
    fn enrollment_token_round_trip() {
        let t = MeshEnrollmentToken {
            token: "AAAAAAAA-BBBBBBBB-CCCC".into(),
            peer_name: "iPhone".into(),
            roles: vec![MeshPeerRole::Member],
            mesh_ip: "100.64.0.5".into(),
            expires_at: 1_700_000_600_000,
        };
        let json = serde_json::to_string(&t).unwrap();
        assert!(json.contains("\"peerName\":\"iPhone\""));
        assert!(json.contains("\"meshIp\":\"100.64.0.5\""));
        let restored: MeshEnrollmentToken = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.token, t.token);
    }

    #[tokio::test]
    async fn stub_engine_state_returns_stopped() {
        let result = mesh_get_engine_state().await;
        assert!(result.success);
        let state = result.data.unwrap();
        assert!(matches!(state.status, MeshEngineStatus::Stopped));
        assert_eq!(state.pid, 0);
    }

    #[tokio::test]
    async fn stub_start_returns_not_implemented_err() {
        let result = mesh_start().await;
        assert!(!result.success);
        assert!(result.error.unwrap().contains("尚未集成"));
    }

    #[tokio::test]
    async fn stub_list_peers_returns_empty() {
        let result = mesh_list_peers().await;
        assert!(result.success);
        assert!(result.data.unwrap().is_empty());
    }
}
